x/incentive/keeper/adapters/hard_borrow/adapter_test.go

```diff
@@ -0,0 +1,285 @@
+}
+
+func (suite *HardBorrowAdapterTestSuite) SetupTest() {
+	config := sdk.GetConfig()
+	app.SetBech32AddressPrefixes(config)
+
+	_, suite.addrs = app.GeneratePrivKeyAddressPairs(5)
+
+	suite.genesisTime = time.Date(2020, 12, 15, 14, 0, 0, 0, time.UTC)
+	suite.app = app.NewTestApp()
+
+	suite.ctx = suite.app.NewContext(true, tmprototypes.Header{Time: suite.genesisTime})
+
+	suite.denomA = "usdx"
+
+	err := suite.app.FundAccount(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(1000000000000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = suite.app.FundAccount(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(1000000000000)),
+		),
+	)
+	suite.NoError(err)
+
+	hardGs := hardtypes.NewGenesisState(
+		hardtypes.NewParams(
+			hardtypes.MoneyMarkets{
+				hardtypes.NewMoneyMarket(
+					suite.denomA,
+					hardtypes.NewBorrowLimit(
+						true,
+						sdk.NewDec(500000000000),
+						sdk.MustNewDecFromStr("0.5"),
+					),
+					"ukava:usd:30",
+					sdk.NewInt(1000000),
+					hardtypes.NewInterestRateModel(
+						sdk.MustNewDecFromStr("0"),
+						sdk.MustNewDecFromStr("0.05"),
+						sdk.MustNewDecFromStr("0.8"),
+						sdk.NewDec(5),
+					),
+					sdk.MustNewDecFromStr("0.025"),
+					sdk.MustNewDecFromStr("0.02"),
+				),
+			},
+			sdk.ZeroDec(),
+		),
+		hardtypes.DefaultAccumulationTimes,
+		nil,
+		nil,
+		sdk.NewCoins(),
+		sdk.NewCoins(),
+		sdk.NewCoins(),
+	)
+
+	pricefeedGs := pricefeedtypes.NewGenesisState(
+		pricefeedtypes.NewParams(
+			[]pricefeedtypes.Market{
+				pricefeedtypes.NewMarket(
+					"ukava:usd:30",
+					"ukava",
+					"usd",
+					nil,
+					true,
+				),
+			},
+		),
+		[]pricefeedtypes.PostedPrice{
+			pricefeedtypes.NewPostedPrice(
+				"ukava:usd:30",
+				suite.addrs[0],
+				sdk.MustNewDecFromStr("1.5"),
+				suite.ctx.BlockTime().Add(time.Hour),
+			),
+		},
+	)
+
+	cdc := suite.app.AppCodec()
+	suite.app.InitializeFromGenesisStates(
+		app.GenesisState{
+			hardtypes.ModuleName:      cdc.MustMarshalJSON(&hardGs),
+			pricefeedtypes.ModuleName: cdc.MustMarshalJSON(&pricefeedGs),
+		},
+	)
+}
+
+func (suite *HardBorrowAdapterTestSuite) TestHardAdapter_OwnerSharesBySource() {
+	hardKeeper := suite.app.GetHardKeeper()
+
+	// Need some deposits in order to borrow
+	err := hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Deposit(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100000)),
+		),
+	)
+	suite.NoError(err)
+
+	// Actual borrows now
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[0],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(100)),
+		),
+	)
+	suite.NoError(err)
+
+	err = hardKeeper.Borrow(
+		suite.ctx,
+		suite.addrs[1],
+		sdk.NewCoins(
+			sdk.NewCoin(suite.denomA, sdk.NewInt(250)),
+		),
+	)
+	suite.NoError(err)
+
```

Comment:
I think it's worth testing the case where `NormalizedBorrow` differs from the borrow amount. If we don't use `NormalizedBorrow` in the adapter the tests won't catch it.

This case happens when the borrow.Index is greater than 1. Which happens when a borrow is modified after interest has accrued. Bit like how a claim's indexes increase when Sync is called after the global indexes have increased.

Should be able to trigger it by:
- borrowing some amount
- stepping through a few blocks to accrue interest on the borrow (interest accrues in hard begin blocker)
- modifying the borrow (borrow or repay) to sync it, therby updating the `Index` field